### PR TITLE
Module 12

### PR DESCRIPTION
**Pull request by MadhuRoy-git**

### CODE REVIEW

File: .travis.yml
Comment (JackMeadDev) [inline .travis.yml]: Don't forget to change this back once this is done

File: main.tf
Comment (JackMeadDev) [inline main.tf]: Here you're not setting any of your other environment variables which are set on your app (like your client id, client secret etc.) - are you manually adding them in via the Portal? If so we should be also specifying those here & registering them as input variables - we can either provide defaults if they're not sensitive, or pass them through if they are

File: .travis.yml
Comment (JackMeadDev) [inline .travis.yml]: Here I'd expect to see some variables passed through like `-var 'CLIENT_ID=$CLIENT_ID'` - not having these suggests that you're either storing sensitive values in your terraform setup, or manually adding those via the Azure Portal

File: outputs.tf
Comment (JackMeadDev) [inline outputs.tf:5]: Worth noting that we don't want this output to be shared, so we should mark it [sensitive](https://www.terraform.io/docs/language/values/outputs.html#sensitive-suppressing-values-in-cli-output) here to avoid it leaking during our pipelines, otherwise we expose the opportunity for anyone to trigger an application deployment

File: variables.tf
Comment (JackMeadDev) [inline variables.tf:11]: I don't think this is used anywhere yet

File: README.md
Comment (JackMeadDev) [inline README.md:82]: Nice, it would be good to also mention how a new developer could work with terraform if they needed to

File: .travis.yml
Comment (JackMeadDev) [inline .travis.yml]: Worth noting that this is erroring in your Travis build currently - you might find it worthwhile adding a `--fail` flag to curl to ensure that it returns an error code if the status code it receives isn't a success.
I can't see a clear reason for that error here, but it could be because of the special characters; I'd suggest trying either  manually adding a `\` to escape the $ in your cd_webhook output, or moving this command into a bash script that's called here as that can subtly change the behaviour

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
Nice work, I can see that Terraform is successfully managing your infrastructure in Travis, and I was able to visit your app successfully, but there are a couple of issues that we should fix up before merging:
* I think you're configuring most of your environment variables manually at the moment - but we should be doing getting Terraform to do that in the app_settings block (and using input variables appropriately)
* Your webhook deployment is currently failing (note the response logged to Travis)

File: .travis.yml
Comment (MadhuRoy-git) [reply .travis.yml]: Hi Jack,
I have made the changes and pushed to my branch. The Travis build (deploy step) fails with this error :22.38s$ terraform apply -var 'CLIENT_ID=$CLIENT_ID' -var 'CLIENT_SECRET=$CLIENT_SECRET' -var 'DOCKER_PASSWORD=$DOCKER_PASSWORD' -auto-approve
dpl_0
1.48s$ rvm $(travis_internal_ruby) --fuzzy do ruby -S gem install dpl
dpl.1
Installing deploy dependencies
dpl.2
Preparing deploy
dpl.3
Deploying application
HEAD detached at a0a01a4
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .terraform.lock.hcl
no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (60540db2bc3083d1e69ef847a81816b47de901d7)
Script failed with status 22
failed to deploy

Comment (MadhuRoy-git) [reply .travis.yml]: I am not sure what is causing this. Could you please let me know ?

File: main.tf
Comment (JackMeadDev) [inline main.tf]: Here you're using a hardcoded string (`"Board123"`) rather than the variable, so the BOARD_ID variable never overrides this

File: variables.tf
Comment (JackMeadDev) [inline variables.tf]: Why do you need this connection string input variable? You're already piecing together the URL from the relevant infrastructure inside `main.tf`

File: variables.tf
Comment (JackMeadDev) [reply variables.tf:11]: This still isn't used anywhere; you've added more variables but you're not actually passing them through to your application inside the `app_settings` block

File: main.tf
Comment (JackMeadDev) [inline main.tf]: This still isn't using some of your variables (like CLIENT_ID, CLIENT_SECRET). I'd expect this to look more like..

```
    app_settings = {
    "BOARD_ID" = var.BOARD_ID
    "DOCKER_REGISTRY_SERVER_URL" = "https://index.docker.io"
    "DOCKER_USER" = var.DOCKER_USER
    "CLIENT_ID" = var.CLIENT_ID
...
```

File: variables.tf
Comment (JackMeadDev) [inline variables.tf]: It would be worthwhile marking this as `sensitive` as well, to help avoid it being leaked by the console, e.g.

```suggestion
variable "CLIENT_SECRET" {
    sensitive   = true
    description = "The Github Client Secret for the Terraform Azure app"
}
```

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
Nice work getting the Travis pipeline working, but your terraform setup still isn't managing most of the environment variables.

In order to pass this, we need to:
* Use the input variables you've specified in `variables.tf` in your app_settings block

Not required for passing, but worthwhile:
* Mark any input/output variables that are secrets as `sensitive` (see example below)
* Add some docs to your Readme explaining how terraform is & can be used

**Review by JackMeadDev — APPROVED** [review]
Thanks for making those changes, I can see that you are now using the terraform input variables to set your Azure settings, and that Travis is still passing (and thanks for adding the _sensitive_ setting to the appropriate input/outputs too). I'm happy for you to merge this when you're ready

**Pull request merged**
